client/connect: ConsulProxy LocalServicePort/Address

tgross opened the pull request:
Fixes #6337

Without a `LocalServicePort`, Connect services will try to use the mapped port even when delivering traffic locally. A user can override this behavior by pinning the port value in the `service` stanza but this prevents us from using the Consul service name to reach the service.

This commits configures the Consul proxy with its `LocalServicePort` and `LocalServiceAddress` fields.

tgross commented on jobspec/test-fixtures/tg-network.hcl:
```diff
@@ -1,12 +1,15 @@
 job "foo" {
   datacenters = ["dc1"]
+
```
I used `hclfmt` to clean up my own changes and it turns out this file was dirty. Apologies for the review noise.

tgross commented:
Would it be at all valuable to expose these as a `NOMAD_*` variable to the tasks?

schmichael commented on command/agent/consul/group_test.go:
```diff
@@ -65,7 +65,10 @@ func TestConsul_Connect(t *testing.T) {
-				SidecarService: &structs.ConsulSidecarService{},
+				SidecarService: &structs.ConsulSidecarService{
+					Proxy: &structs.ConsulProxy{
+						LocalServicePort: 9000},
```
```suggestion
						LocalServicePort: 9000,
					},
```

schmichael commented on nomad/structs/services.go:
```diff
@@ -775,6 +775,13 @@ func (t *SidecarTask) MergeIntoTask(task *Task) {
 // ConsulProxy represents a Consul Connect sidecar proxy jobspec stanza.
 type ConsulProxy struct {
+
+	// LocalServiceAddress ... TODO(tgross)
```
```suggestion
	// LocalServiceAddress is the address the local service is binds to. Usually 127.0.0.1 it is useful to customize in clusters with mixed Connect and non-Connect services.
```

schmichael commented on nomad/structs/services.go:
```diff
@@ -775,6 +775,13 @@ func (t *SidecarTask) MergeIntoTask(task *Task) {
+	// LocalServiceAddress ... TODO(tgross)
+	LocalServiceAddress string
+
+	// LocalServicePort ... TODO(tgross)
```
```suggestion
	// LocalServicePort is the port the local service is binds to. Usually the same as the parent service's port, it is useful to customize in clusters with mixed Connect and non-Connect services.
```

tgross commented:
@schmichael thanks for catching those TODOs and suggesting some good language for them! I've addressed your comments, including adding docs [here](https://github.com/hashicorp/nomad/pull/6358/files#diff-92317c1573f1e4d7fbe559aa956896ba)
